### PR TITLE
Fix infinite loop when zooming on empty bins

### LINKED ISSUE #1527: Freeze when zooming 2D plot of binned data
In the plot generated by this: (taken from https://scipp.github.io/user-guide/binned-data/filtering.html)
```.py
import numpy as np
import scipp as sc
from scipp.plot import plot

np.random.seed(1)
tof_max = 10
width = tof_max/20
sizes = 4*np.array([7, 3, 3, 5])
size = np.sum(sizes)

table = sc.DataArray(data=sc.Variable(dims=['event'],
                     values=np.random.normal(0, 1, size)),
                     coords={'tof': sc.Variable(dims=['event'],
                                                values=np.arange(0., size))})
end = sc.Variable(dims=['x'], values=np.cumsum(sizes))
begin = end.copy()
begin.values -= sizes
events = sc.DataArray(
    data=sc.bins(begin=begin, end=end, dim='event', data=table),
    coords={'x': sc.Variable(dims=['x'], values=np.linspace(0, 1, num=4))})

tof_bins = sc.Variable(dims=['tof'], values=np.linspace(0, tof_max, num=100))
binned_events = sc.bin(events.bins, [tof_bins])
plot(binned_events)
```
When I zoom in, the plot freezes and I can't execute any cells.
The zoom rectangle stays on screen:
frozen_plot

There are no messages printed to terminal. The kernel seems be be busy, it uses 100% of a CPU core.

This does not always happen immediately, it depends on where I zoom in. But when zooming sufficiently deep, it always seems to happen.

I suspect that this is an issue with `transform` and/or `MultiIndex` but I haven't dug into this, so this is just an intuition. Could be related to having empty bins and  #1479

### PR DESCRIPTION
**Pull request by jl-wynen**
Fixed #1527 

The problem is more general, it appears with all operations on binned variables with more than one outer dimension where in all operands at least one outer dimension has extent 0.

### CODE REVIEW

File: core/test/multi_index_test.cpp [hunk 1]
```diff
@@ -287,3 +287,10 @@ TEST_F(MultiIndexTest, two_1d_arrays_of_1d_buckets_bucket_size_mismatch) {
                                   {0, 1, 2, 3, 4, 5, 6}, {0, 1, 2, 3, 4, 5, 6}),
                except::BucketError);
 }
+
+TEST_F(MultiIndexTest, 2d_empty_dims_array_of_1d_buckets) {
+  const Dim dim = Dim::Row;
+  Dimensions buf{dim, 0}; // 1d cut into dims=0x0 sections
```
Comment (SimonHeybrock) [inline core/test/multi_index_test.cpp:293]:
```suggestion
TEST_F(MultiIndexTest, 1d_empty_dims_array_of_1d_buckets) {
  const Dim dim = Dim::Row;
  Dimensions buf{dim, 0}; // 1d cut into dims=0 sections
```

File: dataset/test/bin_test.cpp [hunk 2]
```diff
@@ -89,6 +90,16 @@ TEST_F(DataArrayBinTest, 2d) {
   EXPECT_EQ(bin(bin(table, {edges_x}), {edges_y}), bucketed);
 }
 
+TEST_F(DataArrayBinTest, operations_on_empty) {
+  const Variable indices = makeVariable<std::pair<scipp::index, scipp::index>>(
+      Dimensions{{Dim::Event, 0}, {Dim::Y, 0}}, Values{});
```
Comment (SimonHeybrock) [inline dataset/test/bin_test.cpp:95]: It is confusing to have `Dim::Event` as both an external dimension as well as the internal bin dimension, use something else here?

File: dataset/test/bin_test.cpp [hunk 3]
```diff
@@ -89,6 +90,16 @@ TEST_F(DataArrayBinTest, 2d) {
...
+  const Variable binned = make_bins(indices, Dim::Event, data);
+
+  EXPECT_EQ(abs(binned), binned);
+  EXPECT_EQ(binned, binned * binned);
+  EXPECT_EQ(binned, binned * (2 * units::one));
```
Comment (SimonHeybrock) [inline dataset/test/bin_test.cpp:100]: Not calling `bin`, is this intentional? If so, this test should go into `bins_test.cpp`?

Comment (jl-wynen) [reply dataset/test/bin_test.cpp:100]: Intentional because the version with `bin` is more complex than necessary to track this bug.
-> Moved

**Review by SimonHeybrock — APPROVED** [review]

**Pull request merged**